src/controller/adc.c

```diff
@@ -42,55 +40,88 @@ void adc_init (void)
-    while(TIM3_GetCounter() < ui16_counter) ; // delay ~10ms
+    
+    // delay ~10ms
+    while(TIM3_GetCounter() < ui16_counter);
+    
+    // trigger ADC conversion on all channels (scan conversion, buffered)
     adc_trigger();
-    while(!ADC1_GetFlagStatus(ADC1_FLAG_EOC)) ; // wait for end of conversion
+    
+    // wait for end of conversion
+    while(!ADC1_GetFlagStatus(ADC1_FLAG_EOC));
   }
 
-  // read and average a few values of ADC battery current
-  ui16_adc_battery_current_offset = 0;
-  for(ui8_i = 0; ui8_i < 16; ++ui8_i)
+
+
+  // ADC battery current initialization //////////////////////////////////////////////////////////////////////
+  
+  uint16_t ui16_adc_battery_current_offset = 0;
+
+  // calibrate battery current
+  for (ui8_i = 0; ui8_i < 16; ++ui8_i)
   {
     ui16_counter = TIM3_GetCounter() + 10; // delay ~10ms
-    while(TIM3_GetCounter() < ui16_counter) ; // delay ~10ms
+    
+    // delay ~10ms
+    while(TIM3_GetCounter() < ui16_counter);
+    
+    // trigger ADC conversion on all channels (scan conversion, buffered)
     adc_trigger();
-    while(!ADC1_GetFlagStatus(ADC1_FLAG_EOC)) ; // wait for end of conversion
+    
+    // wait for end of conversion
+    while(!ADC1_GetFlagStatus(ADC1_FLAG_EOC));
+    
     ui16_adc_battery_current_offset += UI8_ADC_BATTERY_CURRENT;
   }
+  
+  // calculate and set current offset
   ui16_adc_battery_current_offset >>= 4;
   ui8_adc_battery_current_offset = ui16_adc_battery_current_offset >> 2;
   ui8_g_adc_motor_phase_current_offset = ui8_adc_battery_current_offset;
 
-  // read and average a few values of ADC torque sensor
-  ui16_adc_torque_sensor_offset = 0;
-  for(ui8_i = 0; ui8_i < 16; ++ui8_i)
+
+
+  // ADC torque sensor initialization //////////////////////////////////////////////////////////////////////
+  
+  #define ADC_TORQUE_SENSOR_OFFSET_ADJUSTMENT   6
+  
+  uint16_t ui16_adc_pedal_torque_offset = 0;
+  
+  // calibrate torque sensor
+  for (ui8_i = 0; ui8_i < 16; ++ui8_i)
   {
     ui16_counter = TIM3_GetCounter() + 10; // delay ~10ms
-    while(TIM3_GetCounter() < ui16_counter) ; // delay ~10ms
-    adc_trigger ();
-    while(!ADC1_GetFlagStatus(ADC1_FLAG_EOC)) ; // wait for end of conversion
-    ui16_adc_torque_sensor_offset += UI8_ADC_TORQUE_SENSOR;
+    
+    while (TIM3_GetCounter() < ui16_counter) ; // delay ~10ms
+    
+    // trigger ADC conversion on all channels (scan conversion, buffered)
+    adc_trigger();
+    
+    // wait for end of conversion
+    while (!ADC1_GetFlagStatus(ADC1_FLAG_EOC));
+    
+    ui16_adc_pedal_torque_offset += UI8_ADC_TORQUE_SENSOR;
   }
-  ui16_adc_torque_sensor_offset >>= 4;
-
-  ui8_g_adc_torque_sensor_min_value = ((uint8_t) ui16_adc_torque_sensor_offset) + ADC_TORQUE_SENSOR_THRESHOLD;
-  ui8_g_adc_torque_sensor_max_value = ui8_g_adc_torque_sensor_min_value + 32;
+  
+  // calculate and set torque offset
+  ui8_adc_pedal_torque_offset = (ui16_adc_pedal_torque_offset >> 4) + ADC_TORQUE_SENSOR_OFFSET_ADJUSTMENT;
```

Comment:
Is the compiler not smart enough to see that "x/16" can be solved by a bitshift?
/16 would better convey the meaning, and mentally link up with the for-loop-counter above :/

Reply:
I think no, this SDCC compiler is not good in many things for the tests I did. I had to optimize and verified on oscilloscope that this kind of optimizations really make difference.